controllers/workloads/console/controller.go

```diff
@@ -117,6 +135,57 @@ func (r *ConsoleReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manag
 		)
 }
 
+func (r *ConsoleReconciler) createOrUpdateUserRbac(logger logr.Logger, ctx context.Context, tpl *workloadsv1alpha1.ConsoleTemplate, req ctrl.Request, csl *workloadsv1alpha1.Console, authorisation *workloadsv1alpha1.ConsoleAuthorisation) error {
+
+	// Create or update the user role
+	role := buildUserRole(req.NamespacedName, csl.Status.PodName)
+	if err := r.createOrUpdate(ctx, logger, csl, role, Role, recutil.RoleDiff); err != nil {
+		return err
+	}
+
+	// Create or update the directory role binding
+	subjects := append(
+		tpl.Spec.AdditionalAttachSubjects,
+		rbacv1.Subject{Kind: "User", Name: csl.Spec.User},
+	)
+	// Append all the authorising users to allow them to attach
+	if authorisation != nil {
+		subjects = append(subjects, authorisation.Spec.Authorisations...)
+	}
+
+	drb := buildUserDirectoryRoleBinding(req.NamespacedName, role, subjects)
+	if err := r.createOrUpdate(ctx, logger, csl, drb, DirectoryRoleBinding, recutil.DirectoryRoleBindingDiff); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *ConsoleReconciler) createOrUpdateServiceRbac(logger logr.Logger, ctx context.Context, tpl *workloadsv1alpha1.ConsoleTemplate, req ctrl.Request, csl *workloadsv1alpha1.Console, authorisation *workloadsv1alpha1.ConsoleAuthorisation) error {
+
+	// We already create roles and directory rolebindings with the same name as
+	// the console to provide permissions on the job/pod. Therefore for the name
+	// of these objects, suffix the console name with '-svc'.
+	rbacName := types.NamespacedName{
+		Name:      fmt.Sprintf("%s-%s", req.Name, "svc"),
+		Namespace: req.Namespace,
+	}
+
+	// Create or update the service role
+	role := buildServiceRole(rbacName, csl.Status.PodName)
+	if err := r.createOrUpdate(ctx, logger, csl, role, Role, recutil.RoleDiff); err != nil {
+		return err
+	}
+
+	// Create or update the service rolebinding
+	drb := buildServiceRoleBinding(rbacName, role, tpl.Spec.Template.Spec.ServiceAccountName)
```

Comment:
Is `ServiceAccountName` guaranteed to be non-empty?

Reply:
I think so!! Do you know if there's a case where it might be empty? Wouldn't a service always have an associated service account that the console would run under?

Reply:
I have now handled the case of a missing service account name!